python-oauth-web-app/app.py

```diff
@@ -0,0 +1,300 @@
+            get_db()
+            .execute("SELECT * FROM oauth_session WHERE did = ?", (user_did,))
+            .fetchone()
+        )
+
+
+def login_required(view):
+    @functools.wraps(view)
+    def wrapped_view(**kwargs):
+        if g.user is None:
+            return redirect("/oauth/login")
+
+        return view(**kwargs)
+
+    return wrapped_view
+
+
+@app.errorhandler(500)
+def internal_server_error(e):
+    return render_template("error.html", status_code=500, err=e), 500
+
+
+@app.errorhandler(400)
+def bad_request_error(e):
+    return render_template("error.html", status_code=400, err=e), 400
+
+
+@app.route("/")
+def homepage():
+    return render_template("home.html")
+
+
+@app.route("/oauth/client-metadata.json")
+def oauth_client_metadata():
+    app_url = request.url_root.replace("http://", "https://")
+    client_id = f"{app_url}oauth/client-metadata.json"
+
+    return jsonify(
+        {
+            # simply using the full request URL for the client_id
+            "client_id": client_id,
+            "dpop_bound_access_tokens": True,
+            "application_type": "web",
+            "redirect_uris": [f"{app_url}oauth/authorize"],
+            "client_uri": app_url,
+            "subject_type": "public",
+            "grant_types": ["authorization_code", "refresh_token"],
+            "response_types": ["code"],  # TODO: "code id_token"?
+            "scope": "openid profile offline_access",
+            "client_name": "atproto OAuth Flask Backend Demo",
+            "token_endpoint_auth_method": "none",
+            "jwks_uri": f"{app_url}oauth/jwks.json",
+            # "jwks": { #    "keys": [pub_jwk], #},
+        }
+    )
+
+
+@app.route("/oauth/jwks.json")
+def oauth_jwks():
+    return jsonify(
+        {
+            "keys": [pub_jwk],
+        }
+    )
+
+
+@app.route("/oauth/login", methods=("GET", "POST"))
+def oauth_login():
+    if request.method != "POST":
+        return render_template("login.html")
+
+    # resolve handle and DID document, using atproto identity helpers
+    username = request.form["username"]
+    if not valid_handle(username):
+        flash("Invalid Handle")
+        return render_template("login.html"), 400
+    did = resolve_handle(username)
+    if not did:
+        flash("Handle Not Found")
+        return render_template("login.html"), 400
+    if not valid_did(did):
+        flash("Handle resolved to invalid DID")
+        return render_template("login.html"), 400
+    did_doc = resolve_did(did)
+    if not did_doc:
+        flash("DID Not Found: " + did)
+        return render_template("login.html"), 400
+
+    # fetch PDS OAuth metadata
+    # IMPORTANT: PDS endpoint URL is untrusted input, SSRF mitigations are needed
+    pds_url = pds_endpoint(did_doc)
+    print(f"PDS: {pds_url}")
+    assert is_safe_url(pds_url)
+    with hardened_http.get_session() as sess:
+        resp = sess.get(f"{pds_url}/.well-known/oauth-protected-resource")
+    resp.raise_for_status()
```

Comment:
Note that specs require a `200` response code and not just a `< 300` response code for most of the calls.

Reply:
added an assertion